groovy-language-server/src/main/java/com/palantir/ls/util/SourceWriter.java

```diff
@@ -0,0 +1,167 @@
+                    newFileContents.append(line.substring(0, Math.min(start.getCharacter(), line.length())));
+                }
+                if (end.getLine() == lineNum) {
+                    // Single line range
+                    newFileContents.append(sortedChanges.get(sortedChangesIndex).getText());
+                    // Add part of line after our change
+                    if (end.getCharacter() + 1 < line.length()) {
+                        newFileContents.append(line.substring(Math.max(end.getCharacter() + 1, 0)));
+                    }
+                    newFileContents.append("\n");
+                } else {
+                    // Multiline range
+                    lineNum = handleMultilineRange(file, newFileContents, lineNum + 1,
+                                    sortedChanges.get(sortedChangesIndex));
+                }
+                ++sortedChangesIndex;
+            } else {
+                newFileContents.append(line).append("\n");
+            }
+
+            ++lineNum;
+        }
+        file.close();
+        return newFileContents.toString();
+    }
+
+    private int handleMultilineRange(BufferedReader file, StringBuilder newFileContents, int currentLineNum,
+            TextDocumentContentChangeEvent change) throws IOException {
+        Position end = change.getRange().getEnd();
+        int lineNum = currentLineNum;
+        String line = null;
+        while ((line = file.readLine()) != null) {
+            if (end.getLine() == lineNum) {
+                newFileContents.append(change.getText());
+                // Add part of line after our change
+                if (end.getCharacter() + 1 < line.length()) {
+                    newFileContents.append(line.substring(Math.max(end.getCharacter() + 1, 0)));
+                }
+                newFileContents.append("\n");
+                return lineNum + 1;
+            }
+            ++lineNum;
+        }
+        // Range exceeds existing number of lines, so just append the change.
+        newFileContents.append(change.getText());
+        return lineNum;
+    }
+
+    /**
+     * Handles saving the accumulated changes in destination back into the original source.
+     */
+    public synchronized void saveChanges() throws IOException {
+        FileUtils.writeStringToFile(source.toFile(), FileUtils.readFileToString(destination.toFile()),
+                StandardCharsets.UTF_8.toString());
+    }
+
+    private synchronized void initialize() throws IOException {
+        if (!destination.toFile().exists() && destination.toFile().isDirectory()) {
+            if (!destination.toFile().mkdirs()) {
+                throw new RuntimeException(
+                        String.format("Could not recreate destination directories: %s", destination.toString()));
+            }
+        }
+        String contents = FileUtils.readFileToString(source.toFile());
+        if (!contents.endsWith("\n")) {
```

Comment:
As discussed in person: https://github.com/palantir/groovy-language-server/issues/65